refactor!: new `aztec` command

benesjan opened the pull request:
The title of the PR is vague because there is a lot of changes in this PR.

Other than tackling these issues here:

Closes https://linear.app/aztec-labs/issue/F-41/replace-aztec-nargo-init-with-aztec-init
Closes https://linear.app/aztec-labs/issue/F-43/replace-aztec-nargo-compile-with-aztec-compile
Closes https://linear.app/aztec-labs/issue/F-44/aztec-compile-help-will-override-the-nargo-compile-help-with-aztec
Closes https://linear.app/aztec-labs/issue/F-48/add-aztec-check-and-aztec-fmt

I also:
- Added `aztec new` command that is like `aztec init` but instead calls `nargo new`. This was necessary as we used `aztec-nargo new` in docs etc. Hence I needed the `aztec new counterpart`.
- I dropped the `aztec-postprocess-contract` directory containing the `transpile_contract_and_gen_vks.sh` script because it now only called the aztec_process command in bb which made this basically a tech debt from time where we forced devs to call this script manually.
- Went to occurrences of "aztec-nargo" and "aztec-postprocess-contract" strings in the codebase and docs and did the relevant updates.
- I replace the default contract template that `nargo init --contract` and `nargo new --contract` commands spit out with:

```
use aztec::macros::aztec;

#[aztec]
contract Main {}
```

This results in the `aztec-nargo` and `aztec-postprocess-contract` being deprecated (and deleted).

Thought it's fine to do all these changes in one PR because the complexity of the changes is low.

benesjan commented:
* **#18417** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18417?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18413** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18413?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18396** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18396?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18398** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18398?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18289** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18289?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18274** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18274?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18265** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18265?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17963** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17963?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18126** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18126?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17792** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17792?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17792?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on aztec-up/bin/aztec:
```diff
@@ -53,7 +53,100 @@ function get_env_vars {
     awk -F"'" '{for(i=2;i<=NF;i+=2) printf $i " "}'
 }
 
+function validate_working_directory {
+  # Validate working directory is under $HOME for Docker containerization
+  if [[ $PWD != ${HOME}* ]]; then
+    >&2 echo "Due to how we containerize our applications, we require your working directory to be somewhere within $HOME."
+    exit 1
+  fi
+}
```
This has been copied over from `aztec-nargo`. Not super sure why it was needed there. If it will be decided that we will use native `nargo` and not the dockerized one we will most likely be able to drop this.

ludamad replied:
This is needed so long as we use the docker container. We can stop soon (or even just call the user installed bb? as long as its on their system)

ludamad commented on aztec-up/bin/aztec:
```diff
@@ -53,7 +53,100 @@ function get_env_vars {
+    echo "Postprocessing contract..."
+
+    # Run postprocessing script
+    docker run ${TTY_FLAGS:-} \
```
Why even have this? the new way is to just do 
```
noir compile
```
then
```
bb aztec_process
```

ludamad replied:
but I guess a one-shot compile makes sense, so defer to you guys

benesjan replied:
I think it's weird to force devs to be aware of the postprocess step so would go with one-shot compile command here.

Regarding why we use the dockerized `nargo` - that will most likely be dropped in a followup PR.

benesjan commented on aztec-up/bin/aztec:
```diff
@@ -53,11 +53,102 @@ function get_env_vars {
+    ;;
   test)
     shift
-    # Should this just be aztec-test? It's like, a new command that doesn't exist on aztec cli.
-    # Or just make this a first class command on aztec cli?
```
This comment was no longer relevant - now we know we want `aztec test`.

benesjan commented on aztec-up/bin/aztec-postprocess-contract:
```diff
@@ -1,25 +0,0 @@
-#!/usr/bin/env bash
```
`aztec compile` handles this automatically so no need to have a separate command.

benesjan commented on aztec-up/bin/aztec:
```diff
@@ -207,6 +297,266 @@ case ${1:-} in
+    # Parse arguments
+    # Note: I allow devs to use the `aztec` command to also create bin and lib projects because we currently rely
+    # on the dockerized nargo and I am not sure if it's expected they will have a native nargo on path. If we decide
+    # to use native nargo I think we can just hardcode --contract here and not bother with the other options.
```
The implementation of the `new` command is basically the same as `init` but I didn't bother reusing the code here as the complexity is low and this comment might make me change quite a few things here.

Feel like the design choice of allowing devs to also create `bin` and `lib` project with `aztec` command is controversial so please chime in if you disagree with the choice.

benesjan commented on yarn-project/aztec/src/cli/cli.ts:
```diff
@@ -37,10 +37,55 @@ export function injectAztecCommands(program: Command, userLog: LogFn, debugLogge
+      Example:
+        $ aztec lsp                     # starts the LSP server
+
+    preload-crs: Downloads and caches the Common Reference String (CRS) data required for zero-knowledge proofs.
+      Example:
+        $ aztec preload-crs             # preloads CRS data
```
This is unrelated to the PR - it was undocumented here and AI added it here when updating the docs.

benesjan commented on aztec-up/bin/aztec:
```diff
@@ -53,11 +53,205 @@ function get_env_vars {
+    $DOCKER_REPO:$VERSION "$@"
+}
+
+function setup_aztec_contract_project {
```
Added this function and `parse_project_flags_and_build_nargo_args` because the functionality is shared between `nargo new` and `nargo init` commands.

benesjan commented on aztec-up/bin/aztec:
```diff
@@ -53,11 +53,205 @@ function get_env_vars {
+        NARGO_ARGS+=("--name" "$2")
+        shift 2
+        ;;
+      --lib)
+        NARGO_ARGS+=("--lib")
+        HAS_TYPE_FLAG=true
+        shift
+        ;;
+      --bin)
+        NARGO_ARGS+=("--bin")
+        HAS_TYPE_FLAG=true
+        shift
+        ;;
+      --contract)
+        NARGO_ARGS+=("--contract")
+        HAS_TYPE_FLAG=true
+        IS_CONTRACT=true
+        shift
+        ;;
```
I allow devs to use the `aztec` command to also create bin and lib projects because we currently rely on the dockerized nargo and I am not sure if it's expected they will have a native nargo on path. If we decide to use native nargo I think we can just hardcode --contract here and not bother with the other options.

I think this is quite a controversial decision so if the reviewer disagrees LMK.

benesjan commented on aztec-up/bin/aztec:
```diff
@@ -53,11 +53,205 @@ function get_env_vars {
+    # If help is requested, show Aztec-specific info then run nargo compile help and then exit in order to not trigger
+    # transpilation
+    for arg in "$@"; do
+      if [ "$arg" == "--help" ] || [ "$arg" == "-h" ]; then
+        cat << 'EOF'
+Aztec Compile - Compile Aztec Noir contracts
+
+This command compiles Aztec Noir contracts using nargo and then automatically
+postprocesses them to generate Aztec-specific artifacts including:
+  - Transpiled contract artifacts
+  - Verification keys
+
+The compiled contracts will be placed in the target/ directory by default.
+
+AZTEC-SPECIFIC NOTES:
+  - Working directory must be under $HOME due to Docker containerization
+  - Compilation automatically includes contract postprocessing
+  - Use standard nargo compile options (see below)
+
+ENVIRONMENT VARIABLES:
+  AZTEC_PATH    Path to Aztec installation (default: $HOME/.aztec)
+  VERSION       Aztec version to use (default: from $AZTEC_PATH/default_version)
+  DOCKER_REPO   Docker repository (default: aztecprotocol/aztec)
+
+---
+Underlying nargo compile options:
```
Added this as requested by [this issue](https://linear.app/aztec-labs/issue/F-44/aztec-compile-help-will-override-the-nargo-compile-help-with-aztec).

benesjan commented on aztec-up/bin/aztec:
```diff
@@ -207,6 +401,151 @@ case ${1:-} in
+    # Run nargo check
+    run_nargo check "$@"
+    ;;
+  lsp)
+    # TODO: Drop this (and the corresponding docs in cli.ts) in case `nargo` is to be used natively (not containerized).
+
+    # Special-case nargo's LSP command:
+    # 1. include --rm for cleanup
+    # 2. don't specify a user (run as root)
+    # 3. don't specify a workdir
+    validate_working_directory
+
+    docker run --rm -i \
+      --name aztec-nargo-lsp \
+      -v $HOME:$HOME \
+      -e HOME=$HOME \
+      --entrypoint=/usr/src/noir/noir-repo/target/release/nargo \
+      $DOCKER_REPO:$VERSION lsp
+    ;;
```
This is a copy-paste from the original `aztec-nargo` command. Might be dropped in the future if it's decided we don't want to use containerized `nargo`.

benesjan commented on boxes/boxes/react/package.json:
```diff
@@ -6,8 +6,8 @@
-    "compile": "cd src/contracts && ${AZTEC_NARGO:-aztec-nargo} compile --silence-warnings && ${AZTEC_POSTPROCESS_CONTRACT:-aztec-postprocess-contract}",
-    "codegen": "${AZTEC_BUILDER:-aztec} codegen src/contracts/target -o artifacts",
+    "compile": "cd src/contracts && ${AZTEC:-aztec} compile --silence-warnings",
+    "codegen": "${AZTEC:-aztec} codegen src/contracts/target -o artifacts",
```
The codegen command is also exposed on `aztec` directly so I used it here.

Think it might make sense to just drop the builder cli and have the functionality exposed just on the `aztec` command but that's out of scope of this PR.

benesjan commented on boxes/aztec-wrapper.sh:
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
+# Wrapper script for aztec command that ensures unique container names
+# This prevents conflicts when building boxes in parallel
```
@ludamad could you please chime in here whether this solution is fine?

I was getting [this failure](http://ci.aztec-labs.com/1a128fad3ed119d6) when building boxes because the repeated use of the `aztec compile` command by `bootstrap.sh` resulted in the container name collisions.

Do you think we should instead try to somehow reuse the same container?

I am a devops noob and this was produced by AI so feel like this is most likely an incorrect approach.

benesjan replied:
As discussed in DMs - addressed this by exporting the CONTAINER_NAME env var instead of using the wrapper.

benesjan commented on boxes/bootstrap.sh:
```diff
@@ -3,12 +3,9 @@ source $(git rev-parse --show-toplevel)/ci3/source_bootstrap
 
 cmd=${1:-}
 
-export TRANSPILER=$PWD/../avm-transpiler/target/release/avm-transpiler
-export BB=$PWD/../barretenberg/cpp/build/bin/bb
-export NARGO=$PWD/../noir/noir-repo/target/release/nargo
-export AZTEC_NARGO=$PWD/../aztec-up/bin/aztec-nargo
-export AZTEC_POSTPROCESS_CONTRACT=$PWD/../aztec-up/bin/aztec-postprocess-contract
-export AZTEC_BUILDER=$PWD/../yarn-project/builder/aztec-builder-dest
```
These were stale and not used so I dropped it.

benesjan commented on boxes/bootstrap.sh:
```diff
@@ -21,6 +18,9 @@ function build {
   npm_install_deps
 
+  # Set VERSION to latest if not defined (required by the aztec command used by yarn build)
+  export VERSION=${VERSION:-latest}
```
@ludamad also feel like this might be incorrect.

Needed to introduce this here because the `aztec compile` command triggered by `yarn build` on line 25 here calls the [run_nargo function](http://ci.aztec-labs.com/1a128fad3ed119d6) and there the version is expected while the `~/.aztec/default_version` file is not available in CI.

Is there a way to get the CI version of the `aztec` docker image here?

Thank you

benesjan replied:
After rebase this was no longer needed and I dropped it so this not relevant anymore.

nventuro reviewed (approved):
Thanks! Reading bash breaks my brain but the functionality seems to be there. If we find small issues here we can address them as we go.

benesjan committed:
refactor!: new `aztec` command

The title of the PR is vague because there is a lot of changes in this PR.

Other than tackling these issues here:

Closes https://linear.app/aztec-labs/issue/F-41/replace-aztec-nargo-init-with-aztec-init
Closes https://linear.app/aztec-labs/issue/F-43/replace-aztec-nargo-compile-with-aztec-compile
Closes https://linear.app/aztec-labs/issue/F-44/aztec-compile-help-will-override-the-nargo-compile-help-with-aztec
Closes https://linear.app/aztec-labs/issue/F-48/add-aztec-check-and-aztec-fmt

I also:
- Added `aztec new` command that is like `aztec init` but instead calls `nargo new`. This was necessary as we used `aztec-nargo new` in docs etc. Hence I needed the `aztec new counterpart`.
- I dropped the `aztec-postprocess-contract` directory containing the `transpile_contract_and_gen_vks.sh` script because it now only called the aztec_process command in bb which made this basically a tech debt from time where we forced devs to call this script manually.
- Went to occurrences of "aztec-nargo" and "aztec-postprocess-contract" strings in the codebase and docs and did the relevant updates.
- I replace the default contract template that `nargo init --contract` and `nargo new --contract` commands spit out with:

```
use aztec::macros::aztec;

#[aztec]
contract Main {}
```

This results in the `aztec-nargo` and `aztec-postprocess-contract` being deprecated (and deleted).

Thought it's fine to do all these changes in one PR because the complexity of the changes is low.
